main/src/ca/uwaterloo/flix/language/ast/Ast.scala

```diff
@@ -79,6 +80,29 @@ object Ast {
     override def hashCode(): Int = input.hashCode()
 
     override def toString: String = name
+
+
+    /**
+      * Gets a line of text from the source as a string.
+      */
+    def getLine(line: Int): String = {
+      @tailrec
+      def rec(ix: Int, lineStartIx: Int, lineNr: Int): String =
+        if (ix < data.length)
+          if (data(ix) == '\n')
+            if (lineNr < line) rec(ix + 1, ix + 1, lineNr + 1)
+            else sliceString(lineStartIx, ix)
+          else rec(ix + 1, lineStartIx, lineNr)
+        else if (lineNr == line) sliceString(lineStartIx, ix)
+        else ""
+
+      rec(ix = 0, lineStartIx = 0, lineNr = 1)
+    }
+
+    /**
+      * Gets a string from a slice of [[data]] from start(inclusive) to end(exclusive).
+      */
+    private def sliceString(start: Int, end: Int) = new String(data, start, math.max(end - start, 0))
```

Comment:
I suspect this is giving out-of-bounds on some strange behavior in VSCode.

Reply:
I am not sure. Now I suspect isWordChar

Reply:
I think for safety we should add bounds check and throw InternalCompilerException

Reply:
I don't think that's necessary. `sliceString` is only called from `getLine` and is private. These lines are a copy paste from `parboiled2` so we shouldn't have issues with them. They have been in production for years! 

Now I think there may be an issue with a `null` check I removed. I'll look into that now.

Reply:
Update: I inlined `sliceString` so it's not possible to use it unsafely. `getLine` does bounds checking (`if (ix < data.length)`) so it's safe to call even with an empty string and with a line number that's out of bounds 👍